src/theory/quantifiers/bv_inverter.cpp

```diff
@@ -261,24 +262,114 @@ Node BvInverter::getPathToPv(Node lit, Node pv, Node sv, Node pvs,
+Node BvInverter::solve_bv_lit(Node sv,
+                              Node lit,
+                              bool pol,
+                              std::vector<unsigned>& path,
+                              BvInverterModelQuery* m,
+                              BvInverterStatus& status) {
+  Assert(!path.empty());
+
   NodeManager* nm = NodeManager::currentNM();
+  unsigned index, nchildren;
+  Kind k;
+
+  index = path.back();
+  Assert(index < lit.getNumChildren());
+  path.pop_back();
+  k = lit.getKind();
+
+  if (k == NOT) {
+    pol = !pol;
+    lit = lit[index];
+  }
+
+  Assert(k == EQUAL
+       || k == BITVECTOR_COMP
+       || k == BITVECTOR_ULT
+       || k == BITVECTOR_ULTBV
+       || k == BITVECTOR_SLT
+       || k == BITVECTOR_SLTBV);
+
+  Assert((k != EQUAL && k != BITVECTOR_COMP) || pol == true);
+
+  Node sv_t = lit[index];
+  Node t = lit[1-index];
+
+  if (k == BITVECTOR_ULT || k == BITVECTOR_ULTBV) {
+    TypeNode solve_tn = sv_t.getType();
+    Node x = getSolveVariable(solve_tn);
+    Node sc, scr, scl;
+    if (index == 0) {
+      scr = nm->mkNode(k, x, t);
+      if (pol == true) {
+        /* x < t
+         * with side condition:
+         * t != 0  */
+        scl = nm->mkNode(DISTINCT, t, bv::utils::mkZero(bv::utils::getSize(t)));
+        sc = nm->mkNode(IMPLIES, scl, scr);
+      } else {
+        sc = scr;
+      }
+    } else if (index == 1) {
+      scr = nm->mkNode(k, t, x);
+      if (pol == true) {
+        /* t < x
+         * with side condition:
+         * t != 1...1  */
+        scl = nm->mkNode(DISTINCT, t, bv::utils::mkOnes(bv::utils::getSize(t)));
+        sc = nm->mkNode(IMPLIES, scl, scr);
+      } else {
+        sc = scr;
+      }
+    }
+    status.d_conds.push_back(sc);
+    /* t = skv (fresh skolem constant)  */
+    Node skv = getInversionNode(sc, solve_tn);
+    t = skv;
+  } else if (k == BITVECTOR_SLT || k == BITVECTOR_SLTBV) {
+    TypeNode solve_tn = sv_t.getType();
+    Node x = getSolveVariable(solve_tn);
+    Node sc, scr, scl;
+    unsigned w = bv::utils::getSize(t);
+    if (index == 0) {
+      scr = nm->mkNode(k, x, t);
+      if (pol == true) {
+        /* x < t
+         * with side condition:
+         * t != 10...0 */
+        Node min = bv::utils::mkConst(BitVector(w).setBit(w - 1));
+        scl = nm->mkNode(DISTINCT, min, t);
+        sc = nm->mkNode(IMPLIES, scl, scr);
+      } else {
+        sc = scr;
+      }
+    } else if (index == 1) {
+      scr = nm->mkNode(k, t, x);
+      if (pol == true) {
+        /* t < x
+         * with side condition:
+         * t != 01...1  */
+        BitVector bv = BitVector(w).setBit(w - 1);
+        Node max = bv::utils::mkConst(~bv);
+        scl = nm->mkNode(DISTINCT, t, max);
+        sc = nm->mkNode(IMPLIES, scl, scr);
+      } else {
+        sc = scr;
+      }
+    }
+    status.d_conds.push_back(sc);
+    /* t = skv (fresh skolem constant)  */
+    Node skv = getInversionNode(sc, solve_tn);
+    t = skv;
+  }
```

Comment:
Do we need a case for EQUAL here?  Handling EQUAL should depend on pol.  It looks like the above assertion is saying that can't happen, but this surprises me, I think we could get disequalities as input here.

Reply:
See above.

Reply:
Btw, I realized that the handling for LT is incorrect (no step down). I'll fix that.